# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/wr/webrender/src/render_backend.rs
# Commit: 96da9d241051
# Full Hash: 96da9d241051d222bdd693bdb519d058a013b59c
# Author: Doug Thayer <dothayer@mozilla.com>
# Date: 2019-03-23 09:46:24
# Regressor Bug: 1441308
# File Overlap Count: 2
# Description:
#   Bug 1441308 - Core renderroot splitting changes r=kats,sotaro
#   
#   This is a large patch that contains all of the core changes for
#   renderroot splitting.
#   
# ==============================================================================

diff -r 8cefa694f811 -r 96da9d241051 gfx/wr/webrender/src/render_backend.rs
--- a/gfx/wr/webrender/src/render_backend.rs	Fri Mar 22 18:28:31 2019 +0000
+++ b/gfx/wr/webrender/src/render_backend.rs	Fri Mar 22 18:28:42 2019 +0000
@@ -302,13 +302,15 @@
 }
 
 struct Document {
+    // The id of this document
+    id: DocumentId,
     // The latest built scene, usable to build frames.
     // received from the scene builder thread.
     scene: Scene,
 
     // Temporary list of removed pipelines received from the scene builder
     // thread and forwarded to the renderer.
-    removed_pipelines: Vec<PipelineId>,
+    removed_pipelines: Vec<(PipelineId, DocumentId)>,
 
     view: DocumentView,
 
@@ -362,6 +364,7 @@
         default_device_pixel_ratio: f32,
     ) -> Self {
         Document {
+            id,
             scene: Scene::new(),
             removed_pipelines: Vec::new(),
             view: DocumentView {
@@ -565,7 +568,8 @@
     pub fn updated_pipeline_info(&mut self) -> PipelineInfo {
         let removed_pipelines = self.removed_pipelines.take_and_preallocate();
         PipelineInfo {
-            epochs: self.scene.pipeline_epochs.clone(),
+            epochs: self.scene.pipeline_epochs.iter()
+                .map(|(&pipeline_id, &epoch)| ((pipeline_id, self.id), epoch)).collect(),
             removed_pipelines,
         }
     }
@@ -829,7 +833,7 @@
             }
             SceneMsg::RemovePipeline(pipeline_id) => {
                 profile_scope!("RemovePipeline");
-                txn.removed_pipelines.push(pipeline_id);
+                txn.removed_pipelines.push((pipeline_id, document_id));
             }
             SceneMsg::EnableFrameOutput(pipeline_id, enable) => {
                 if enable {
@@ -918,7 +922,7 @@
                     }
                     SceneBuilderResult::ClearNamespace(id) => {
                         self.resource_cache.clear_namespace(id);
-                        self.documents.retain(|doc_id, _doc| doc_id.0 != id);
+                        self.documents.retain(|doc_id, _doc| doc_id.namespace_id != id);
                     }
                     SceneBuilderResult::Stopped => {
                         panic!("We haven't sent a Stop yet, how did we get a Stopped back?");
@@ -1011,7 +1015,7 @@
             }
             ApiMsg::CloneApiByClient(namespace_id) => {
                 assert!(self.namespace_alloc_by_client);
-                debug_assert!(!self.documents.iter().any(|(did, _doc)| did.0 == namespace_id));
+                debug_assert!(!self.documents.iter().any(|(did, _doc)| did.namespace_id == namespace_id));
             }
             ApiMsg::AddDocument(document_id, initial_size, layer) => {
                 let document = Document::new(
@@ -1020,7 +1024,8 @@
                     layer,
                     self.default_device_pixel_ratio,
                 );
-                self.documents.insert(document_id, document);
+                let old = self.documents.insert(document_id, document);
+                debug_assert!(old.is_none());
             }
             ApiMsg::DeleteDocument(document_id) => {
                 self.documents.remove(&document_id);
@@ -1302,7 +1307,7 @@
         // async transforms.
         if requested_frame || has_built_scene {
             if let Some(ref sampler) = self.sampler {
-                frame_ops.append(&mut sampler.sample());
+                frame_ops.append(&mut sampler.sample(document_id));
             }
         }
 
@@ -1627,7 +1632,7 @@
         for (&id, doc) in &mut self.documents {
             debug!("\tdocument {:?}", id);
             if config.bits.contains(CaptureBits::SCENE) {
-                let file_name = format!("scene-{}-{}", (id.0).0, id.1);
+                let file_name = format!("scene-{}-{}", id.namespace_id.0, id.id);
                 config.serialize(&doc.scene, file_name);
             }
             if config.bits.contains(CaptureBits::FRAME) {
@@ -1640,15 +1645,15 @@
                 //TODO: write down doc's pipeline info?
                 // it has `pipeline_epoch_map`,
                 // which may capture necessary details for some cases.
-                let file_name = format!("frame-{}-{}", (id.0).0, id.1);
+                let file_name = format!("frame-{}-{}", id.namespace_id.0, id.id);
                 config.serialize(&rendered_document.frame, file_name);
-                let file_name = format!("clip-scroll-{}-{}", (id.0).0, id.1);
+                let file_name = format!("clip-scroll-{}-{}", id.namespace_id.0, id.id);
                 config.serialize_tree(&doc.clip_scroll_tree, file_name);
-                let file_name = format!("builder-{}-{}", (id.0).0, id.1);
+                let file_name = format!("builder-{}-{}", id.namespace_id.0, id.id);
                 config.serialize(doc.frame_builder.as_ref().unwrap(), file_name);
             }
 
-            let data_stores_name = format!("data-stores-{}-{}", (id.0).0, id.1);
+            let data_stores_name = format!("data-stores-{}-{}", id.namespace_id.0, id.id);
             config.serialize(&doc.data_stores, data_stores_name);
         }
 
@@ -1729,19 +1734,20 @@
 
         for (id, view) in backend.documents {
             debug!("\tdocument {:?}", id);
-            let scene_name = format!("scene-{}-{}", (id.0).0, id.1);
+            let scene_name = format!("scene-{}-{}", id.namespace_id.0, id.id);
             let scene = CaptureConfig::deserialize::<Scene, _>(root, &scene_name)
                 .expect(&format!("Unable to open {}.ron", scene_name));
 
-            let interners_name = format!("interners-{}-{}", (id.0).0, id.1);
+            let interners_name = format!("interners-{}-{}", id.namespace_id.0, id.id);
             let interners = CaptureConfig::deserialize::<Interners, _>(root, &interners_name)
                 .expect(&format!("Unable to open {}.ron", interners_name));
 
-            let data_stores_name = format!("data-stores-{}-{}", (id.0).0, id.1);
+            let data_stores_name = format!("data-stores-{}-{}", id.namespace_id.0, id.id);
             let data_stores = CaptureConfig::deserialize::<DataStores, _>(root, &data_stores_name)
                 .expect(&format!("Unable to open {}.ron", data_stores_name));
 
             let mut doc = Document {
+                id: id,
                 scene: scene.clone(),
                 removed_pipelines: Vec::new(),
                 view: view.clone(),
@@ -1760,7 +1766,7 @@
                 render_task_counters: RenderTaskTreeCounters::new(),
             };
 
-            let frame_name = format!("frame-{}-{}", (id.0).0, id.1);
+            let frame_name = format!("frame-{}-{}", id.namespace_id.0, id.id);
             let frame = CaptureConfig::deserialize::<Frame, _>(root, frame_name);
             let build_frame = match frame {
                 Some(frame) => {